autogpt_platform/backend/backend/integrations/oauth/google.py

```diff
@@ -35,36 +47,74 @@ def get_login_url(self, scopes: list[str], state: str) -> str:
         )
         return authorization_url
 
-    def exchange_code_for_tokens(self, code: str) -> OAuth2Credentials:
-        flow = self._setup_oauth_flow(None)
+    def exchange_code_for_tokens(
+        self, code: str, scopes: list[str]
+    ) -> OAuth2Credentials:
+        logger.debug(f"Exchanging code for tokens with scopes: {scopes}")
+
+        # Use the scopes from the initial request
+        flow = self._setup_oauth_flow(scopes)
         flow.redirect_uri = self.redirect_uri
-        flow.fetch_token(code=code)
 
-        google_creds = flow.credentials
-        username = self._request_email(google_creds)
+        logger.debug("Fetching token from Google")
 
-        # Google's OAuth library is poorly typed so we need some of these:
-        assert google_creds.token
-        assert google_creds.refresh_token
-        assert google_creds.expiry
-        assert google_creds.scopes
-        return OAuth2Credentials(
-            provider=self.PROVIDER_NAME,
-            title=None,
-            username=username,
-            access_token=SecretStr(google_creds.token),
-            refresh_token=SecretStr(google_creds.refresh_token),
-            access_token_expires_at=int(google_creds.expiry.timestamp()),
-            refresh_token_expires_at=None,
-            scopes=google_creds.scopes,
-        )
+        try:
+            # Disable scope check in fetch_token
+            flow.oauth2session.scope = None
+            token = flow.fetch_token(code=code)
+            logger.debug("Token fetched successfully")
+
+            # Get the actual scopes granted by Google
+            granted_scopes = token.get("scope", [])
+            if isinstance(granted_scopes, str):
+                granted_scopes = granted_scopes.split()
```

Comment:
does Google *sometimes* return a comma-separated string but not always?

Reply:
its for when we get scopes back. Otherwise scopes is an array

Reply:
then I would instead do:
```py
granted_scopes_str = token.get("scope", "")
granted_scopes = granted_scopes_str.split() if granted_scopes_str else []
```
to be clear about what's coming from where.

Reply:
Also, when is `token["scope"]` non-existent?

Reply:
done, it never SHOULD be but possibly COULD be. Their typing leaves much to be desired